src/solvers/iterators.jl

```diff
@@ -1,100 +1,158 @@
-region_kwargs(R::RegionIterator) = current_region_plan(R)[2]
-function previous_region(R::RegionIterator)
-  R.which_region==1 ? nothing : R.region_plan[R.which_region - 1][1]
+
+function current_region(R::RegionIterator)
+  region, _ = current_region_plan(R)
+  return region
 end
-function next_region(R::RegionIterator)
-  R.which_region==length(R.region_plan) ? nothing : R.region_plan[R.which_region + 1][1]
+
+function current_region_kwargs(R::RegionIterator)
+  _, kwargs = current_region_plan(R)
+  return kwargs
+end
+
+function previous_region(R::RegionIterator)
+  state(R) <= 1 && return nothing
+  prev, _ = R.region_plan[R.which_region - 1]
+  return prev
 end
-is_last_region(R::RegionIterator) = isnothing(next_region(R))
 
-function Base.iterate(R::RegionIterator, which=1)
-  R.which_region = which
-  region_plan_state = iterate(R.region_plan, which)
-  isnothing(region_plan_state) && return nothing
-  (current_region, region_kwargs), next = region_plan_state
-  R.problem = region_step(problem(R), R; region_kwargs...)
-  return R, next
+function next_region(R::RegionIterator)
+  is_last_region(R) && return nothing
+  next, _ = R.region_plan[R.which_region + 1]
+  return next
 end
+is_last_region(R::RegionIterator) = length(R) === state(R)
 
 #
 # Functions associated with RegionIterator
 #
 
-function region_iterator(problem; sweep_kwargs...)
-  return RegionIterator(; problem, region_plan=region_plan(problem; sweep_kwargs...))
+function compute!(R::RegionIterator)
+  region_kwargs = current_region_kwargs(R)
+  R.problem = region_step(R; region_kwargs...)
+  return R
+end
+function increment!(R::RegionIterator)
+  R.which_region += 1
+  return R
+end
+
+function RegionIterator(problem; sweep, sweep_kwargs...)
+  plan = region_plan(problem; sweep, sweep_kwargs...)
+  return RegionIterator(problem, plan, sweep)
 end
 
 function region_step(
-  problem,
-  region_iterator;
-  extract_kwargs=(;),
-  update_kwargs=(;),
-  insert_kwargs=(;),
-  sweep,
-  kws...,
+  region_iterator; extract_kwargs=(;), update_kwargs=(;), insert_kwargs=(;), kws...
 )
-  problem, local_state = extract(problem, region_iterator; extract_kwargs..., sweep, kws...)
-  problem, local_state = update(
-    problem, local_state, region_iterator; update_kwargs..., kws...
-  )
-  problem = insert(problem, local_state, region_iterator; sweep, insert_kwargs..., kws...)
-  return problem
+  prob = problem(region_iterator)
+
+  sweep = region_iterator.sweep
+
+  prob, local_state = extract(prob, region_iterator; extract_kwargs..., sweep, kws...)
+  prob, local_state = update(prob, local_state, region_iterator; update_kwargs..., kws...)
+  prob = insert(prob, local_state, region_iterator; sweep, insert_kwargs..., kws...)
+  return prob
 end
 
 function region_plan(problem; kws...)
   return euler_sweep(state(problem); kws...)
 end
+
+#
+# SweepIterator
+#
+
+mutable struct SweepIterator{Problem} <: AbstractNetworkIterator
+  sweep_kws
```

Comment:
Should we parametrize `SweepIterator` by the type of `sweep_kws`? Or are we expecting it could be modified such that it changes type (though if that is the case we could make it explicit by setting the type parameter to `Any`)?